routing/localchans/manager.go

```diff
@@ -180,10 +247,88 @@ func (r *Manager) UpdatePolicy(newSchema routing.ChannelPolicy,
+func (r *Manager) createEdge(channel *channeldb.OpenChannel,
+	timestamp time.Time) (*models.ChannelEdgeInfo,
+	*models.ChannelEdgePolicy, error) {
+
+	nodeKey1Bytes := r.SelfPub.SerializeCompressed()
+	nodeKey2Bytes := channel.IdentityPub.SerializeCompressed()
+	bitcoinKey1Bytes := channel.LocalChanCfg.MultiSigKey.PubKey.
+		SerializeCompressed()
+	bitcoinKey2Bytes := channel.RemoteChanCfg.MultiSigKey.PubKey.
+		SerializeCompressed()
+	channelFlags := lnwire.ChanUpdateChanFlags(0)
+
+	// Make it such that node_id_1 is the lexicographically-lesser of the
+	// two compressed keys sorted in ascending lexicographic order.
+	if bytes.Compare(nodeKey2Bytes, nodeKey1Bytes) < 0 {
+		nodeKey1Bytes, nodeKey2Bytes = nodeKey2Bytes, nodeKey1Bytes
+		bitcoinKey1Bytes, bitcoinKey2Bytes = bitcoinKey2Bytes,
+			bitcoinKey1Bytes
+		channelFlags = 1
+	}
+
+	var featureBuf bytes.Buffer
+	err := lnwire.NewRawFeatureVector().Encode(&featureBuf)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to encode features: %w",
+			err)
+	}
+
+	info := &models.ChannelEdgeInfo{
+		ChannelID:    channel.ShortChanID().ToUint64(),
+		ChainHash:    channel.ChainHash,
+		Features:     featureBuf.Bytes(),
+		Capacity:     channel.Capacity,
+		ChannelPoint: channel.FundingOutpoint,
+	}
+
+	copy(info.NodeKey1Bytes[:], nodeKey1Bytes)
+	copy(info.NodeKey2Bytes[:], nodeKey2Bytes)
+	copy(info.BitcoinKey1Bytes[:], bitcoinKey1Bytes)
+	copy(info.BitcoinKey2Bytes[:], bitcoinKey2Bytes)
+
+	// Construct a dummy channel edge policy with default values that will
+	// be updated with the new values in the call to processChan below.
+	timeLockDelta := uint16(r.DefaultRoutingPolicy.TimeLockDelta)
+	edge := &models.ChannelEdgePolicy{
+		ChannelID:                 channel.ShortChanID().ToUint64(),
+		LastUpdate:                timestamp,
+		TimeLockDelta:             timeLockDelta,
+		ChannelFlags:              channelFlags,
+		MessageFlags:              lnwire.ChanUpdateRequiredMaxHtlc,
+		FeeBaseMSat:               r.DefaultRoutingPolicy.BaseFee,
+		FeeProportionalMillionths: r.DefaultRoutingPolicy.FeeRate,
+		MinHTLC:                   r.DefaultRoutingPolicy.MinHTLCOut,
+		MaxHTLC:                   r.DefaultRoutingPolicy.MaxHTLC,
+	}
+
+	copy(edge.ToNode[:], channel.IdentityPub.SerializeCompressed())
+
+	return info, edge, nil
+}
+
+// updateEdge updates the given edge with the new schema. The edge parameter may
+// be nil, in that case a new channel policy is returned. In other cases the
+// passed in channel policy is returned after modification.
 func (r *Manager) updateEdge(tx kvdb.RTx, chanPoint wire.OutPoint,
 	edge *models.ChannelEdgePolicy,
-	newSchema routing.ChannelPolicy) error {
+	newSchema routing.ChannelPolicy) (*models.ChannelEdgePolicy, error) {
+
+	channel, err := r.FetchChannel(tx, chanPoint)
+	if err != nil {
+		return nil, err
+	}
+
+	// If due to some unforeseen circumstances the policy doesn't exist,
+	// recreate it here.
+	if edge == nil {
```

Comment:
Ok makes sense, let's do that, and you @JssDWt can run this version of the PR fix the edge which were messed up with the prior version and then return an error here.

Reply:
We shouldn't remove this. Consider the case where the user inputs an invalid value for the channel policy, like a max_htlc_msat that is too large. In that case the edge will be created, but the policy will not. It leaves the system in an inconsistent state.

Reply:
maybe then hide it behind a flag, so we will very likely capture the bug and the node runner can at least fix his situation if he wants to ?

Reply:
One is the edge info and the other is the edge policy, I'm not sure how an invalid policy should lead to the recreation of the edge. Plus when the user inputs an invalid value, it should be caught by the RPC - if not, the right fix is to make sure the policy is validated before updating it, which IIRC has already been handled.

We encountered this issue only because we forgot to add the timestamp when manually creating the edge, which is a bug that should have been fixed and has already been fixed. And we only use the `create_missing_edge` mitigation because the root cause is difficult to find, in other words we are still hiding the bug. If we then decide to go with the `edge == nil` approach, that means we are hiding even more bugs.

Reply:
I considered the case where the user does `lncli updatechanpolicy --create_missing_edge`, but enters an invalid `max_htlc_msat` flag. Currently this happens:
- edge is inserted
- policy is not created due to the validation error

What we could do is swap around these two lines:
https://github.com/lightningnetwork/lnd/pull/8805/files#diff-888432a99a4328c46ba85b63e86398ee6a79fc4fc4da59f7a30a1db88b2ddf91R203
https://github.com/lightningnetwork/lnd/pull/8805/files#diff-888432a99a4328c46ba85b63e86398ee6a79fc4fc4da59f7a30a1db88b2ddf91R216

So the policy is validated before inserting the edge.

Alternatively, do `edge == nil && createMissingEdge` inside that check to recreate the policy.

Reply:
I think we cannot do swapping, because if the edge gets updated successfully but the edge fails to be created we error when `PropagateChanPolicyUpdate`.

What about just validating the edge policy by just calling the updateEdge before (but with a non-pointer as proposed in my other comment) ?

Reply:
OK, sounds good. Next question is what to do with the nil edge policy then. My proposal is to skip processing it, rather than erroring as it was before. We could log a warning. The thing is if you ever have any channel that is missing a policy, any other channel fails to update here. By skipping the update would only fail on that specific channel, but not for other channels.

Reply:
I pushed a commit for tryout. Let me know what you think:
a02b99f765b1cb58a781886c50357192873d922c

Reply:
sorry for going back and forth on this...basically I think we are missing a new things here, maybe it's easier to show a diff,
```diff
diff --git a/routing/localchans/manager.go b/routing/localchans/manager.go
index ba98cdb73..c184c165c 100644
--- a/routing/localchans/manager.go
+++ b/routing/localchans/manager.go
@@ -83,9 +83,7 @@ func (r *Manager) UpdatePolicy(newSchema routing.ChannelPolicy,
 	policiesToUpdate := make(map[wire.OutPoint]models.ForwardingPolicy)
 
 	// NOTE: edge may be nil when this function is called.
-	processChan := func(
-		tx kvdb.RTx,
-		info *models.ChannelEdgeInfo,
+	processChan := func(tx kvdb.RTx, info *models.ChannelEdgeInfo,
 		edge *models.ChannelEdgePolicy) error {
 
 		// If we have a channel filter, and this channel isn't a part
@@ -100,9 +98,17 @@ func (r *Manager) UpdatePolicy(newSchema routing.ChannelPolicy,
 		// for whatever reason the policy does not exist for a channel,
 		// it doesn't affect the updating of other channels.
 		if edge == nil {
-			log.Warnf("Got nil channel edge policy when updating "+
+			log.Errorf("Got nil channel edge policy when updating "+
 				"a channel. Channel point: %v",
 				info.ChannelPoint.String())
+
+			upd := makeFailureItem(
+				info.ChannelPoint,
+				lnrpc.UpdateFailure_UPDATE_FAILURE_NOT_FOUND,
+				"edge not found",
+			)
+			failedUpdates = append(failedUpdates, upd)
+
 			return nil
 		}
 
@@ -184,13 +190,13 @@ func (r *Manager) UpdatePolicy(newSchema routing.ChannelPolicy,
 					"not yet confirmed",
 				))
 
+		// If the edge was not found, but the channel is found, that
+		// means the edge is missing in the graph database and should
+		// be recreated. The edge and policy are created in-memory. The
+		// edge is inserted in createEdge below and the policy will be
+		// added to the graph in the PropagateChanPolicyUpdate call
+		// below.
 		case createMissingEdge:
-			// If the edge was not found, but the channel is found,
-			// that means the edge is missing in the graph database
-			// and should be recreated. The edge and policy are
-			// created in-memory. The edge is inserted in createEdge
-			// below and the policy will be added to the graph in
-			// the PropagateChanPolicyUpdate call below.
 			log.Warnf("Missing edge for active channel (%s) "+
 				"during policy update. Recreating edge with "+
 				"default policy.",
@@ -365,15 +371,6 @@ func (r *Manager) updateEdge(tx kvdb.RTx, chanPoint wire.OutPoint,
 		return nil, err
 	}
 
-	// If due to some unforeseen circumstances the policy doesn't exist,
-	// recreate it here.
-	if edge == nil {
-		_, edge, err = r.createEdge(channel, time.Now())
-		if err != nil {
-			return nil, err
-		}
-	}
-
 	// Update forwarding fee scheme and required time lock delta.
 	edge.FeeBaseMSat = newSchema.BaseFee
 	edge.FeeProportionalMillionths = lnwire.MilliSatoshi(
```

Reply:
Since the `edge` is a param sent to `updateEdge`, I think we can skip the nil check as it would be caught by the test anyway. And other nits and stuff 🙏